rewrite-java/src/main/java/org/openrewrite/java/cleanup/IsBlockLambdaConvertibleVisitor.java

```diff
@@ -0,0 +1,115 @@
+import org.apache.commons.lang3.tuple.Pair;
+import org.openrewrite.Cursor;
+import org.openrewrite.java.JavaIsoVisitor;
+import org.openrewrite.java.tree.J;
+import org.openrewrite.java.tree.JavaType;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+@Value
+@EqualsAndHashCode(callSuper = true)
+public class IsBlockLambdaConvertibleVisitor extends JavaIsoVisitor<AtomicBoolean> {
+
+    Map<Pair<String, JavaType>, Boolean> localVariables = new HashMap<>();
+    Pair<String, JavaType> excludedVariable;
+
+    public static AtomicBoolean isBlockLambdaConvertible(J.Block block, Cursor cursor, J.Identifier _excludedVariable) {
+        IsBlockLambdaConvertibleVisitor isBlockLambdaConvertibleVisitor = new IsBlockLambdaConvertibleVisitor(new ImmutablePair<String, JavaType>(_excludedVariable.getSimpleName(), _excludedVariable.getFieldType()));
+        isBlockLambdaConvertibleVisitor.populateLocalVariables(cursor);
+        return isBlockLambdaConvertibleVisitor.reduce(block, new AtomicBoolean(true));
+    }
+
+    @Override
+    public J.Assignment visitAssignment(J.Assignment assignment, AtomicBoolean isConvertible) {
+        if (!isConvertible.get()) {
+            return assignment;
+        }
+        Pair<String, JavaType> variable = new ImmutablePair<>(((J.Identifier) assignment.getVariable()).getSimpleName(), ((J.Identifier) assignment.getVariable()).getFieldType());
+        if (localVariables.containsKey(variable)) {
+            isConvertible.set(false);
+            return assignment;
+        }
+        return super.visitAssignment(assignment, isConvertible);
+    }
+
+    @Override
+    public J.Identifier visitIdentifier(J.Identifier identifier, AtomicBoolean isConvertible) {
+        if (!isConvertible.get()) {
+            return identifier;
+        }
+        if (Objects.nonNull(identifier.getFieldType())) {
+            Pair<String, JavaType> variable = new ImmutablePair<>(identifier.getSimpleName(), identifier.getFieldType());
+            if (localVariables.containsKey(variable) && !localVariables.get(variable)) {
+                isConvertible.set(false);
+                return identifier;
+            }
+        }
+        return super.visitIdentifier(identifier, isConvertible);
+    }
+
+    @Override
+    public J.Return visitReturn(J.Return _return, AtomicBoolean isConvertible) {
+        if (!isConvertible.get()) {
+            return _return;
+        }
+        isConvertible.set(false);
+        return _return;
+    }
+
+    private void populateLocalVariables(Cursor cursor) {
+        J.MethodDeclaration methodDeclaration = cursor.firstEnclosing(J.MethodDeclaration.class);
+        if (methodDeclaration == null) {
+            throw new IllegalStateException("A method declaration is required in the cursor path.");
+        }
+
+        new JavaIsoVisitor<Map<Pair<String, JavaType>, Boolean>>() {
```

Comment:
I think this visitor needs to stop the recursion at "class boundaries". So when a class declaration, anonymous class or a lambda is encountered, the visitor should not descend into these, as they are not allowed to assign to the variables from the lexical scope in which they are declared, so no need to check them. Additionally, checking them is problematic, because they could introduce a new variable which shades the one you want to check.

In addition to the above comment, the check for reassigned variables should apply to method input parameters. So the following should not be convertible:
```java
class A {
    void method(int x) {
        x = 10;
        Optional<Integer> o = Optional.of(2);
        o.ifPresent((obj) -> {
            System.out.println(x);
        });
    }
}
```

Also, the visitor should follow all other scoping rules. So for example the following code should be convertible, because the x being assigned in the second block is a different variable:
```java
class A {
    void method() {
        {
            int x = 10;
            Optional<Integer> o = Optional.of(2);
            if (o.isPresent()) {
                System.out.println(x);
            }
        }
        {
            int x = 10;
            x = 20;
        }
    }
}
```
or a slightly less esoteric example, but with same principle idea:
```java
class A {
    void method() {
        if (true) {
            int x = 10;
            Optional<Integer> o = Optional.of(2);
            if (o.isPresent()) {
                System.out.println(x);
            }
        } else {
            int x = 10;
            x = 20;
        }
    }
}
```